ql/src/java/org/apache/hadoop/hive/ql/parse/rewrite/CopyOnWriteUpdateRewriter.java

```diff
@@ -0,0 +1,132 @@
+import org.apache.hadoop.hive.ql.parse.rewrite.sql.COWWithClauseBuilder;
+import org.apache.hadoop.hive.ql.parse.rewrite.sql.MultiInsertSqlBuilder;
+import org.apache.hadoop.hive.ql.parse.rewrite.sql.SetClausePatcher;
+import org.apache.hadoop.hive.ql.parse.rewrite.sql.SqlBuilderFactory;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class CopyOnWriteUpdateRewriter implements Rewriter<UpdateStatement> {
+
+  private final HiveConf conf;
+  private final SqlBuilderFactory sqlBuilderFactory;
+  private final COWWithClauseBuilder cowWithClauseBuilder;
+  private final SetClausePatcher setClausePatcher;
+
+
+  public CopyOnWriteUpdateRewriter(HiveConf conf, SqlBuilderFactory sqlBuilderFactory,
+                                   COWWithClauseBuilder cowWithClauseBuilder, SetClausePatcher setClausePatcher) {
+    this.conf = conf;
+    this.sqlBuilderFactory = sqlBuilderFactory;
+    this.cowWithClauseBuilder = cowWithClauseBuilder;
+    this.setClausePatcher = setClausePatcher;
+  }
+
+  @Override
+  public ParseUtils.ReparseResult rewrite(Context context, UpdateStatement updateBlock)
+      throws SemanticException {
+
+    String whereClause = context.getTokenRewriteStream().toString(
+        updateBlock.getWhereTree().getChild(0).getTokenStartIndex(),
+        updateBlock.getWhereTree().getChild(0).getTokenStopIndex());
+    String filePathCol = HiveUtils.unparseIdentifier("FILE__PATH", conf);
+
+    MultiInsertSqlBuilder sqlBuilder = sqlBuilderFactory.createSqlBuilder();
+
+    cowWithClauseBuilder.appendWith(sqlBuilder, filePathCol, whereClause);
+
+    sqlBuilder.append("insert into table ");
+    sqlBuilder.appendTargetTableName();
+    sqlBuilder.appendPartitionColsOfTarget();
+
+    int columnOffset = sqlBuilder.getDeleteValues(Context.Operation.UPDATE).size();
+    sqlBuilder.append(" select ");
+    sqlBuilder.appendAcidSelectColumns(Context.Operation.UPDATE);
+    sqlBuilder.removeLastChar();
+
+    Map<Integer, ASTNode> setColExprs = new HashMap<>(updateBlock.getSetCols().size());
+    List<FieldSchema> nonPartCols = updateBlock.getTargetTable().getCols();
+    for (int i = 0; i < nonPartCols.size(); i++) {
+      sqlBuilder.append(',');
+      String name = nonPartCols.get(i).getName();
+      ASTNode setCol = updateBlock.getSetCols().get(name);
+      String identifier = HiveUtils.unparseIdentifier(name, this.conf);
+      sqlBuilder.append(identifier);
+      sqlBuilder.append(" AS ").append(identifier);
+      if (setCol != null) {
+        // This is one of the columns we're setting, record it's position so we can come back
+        // later and patch it up.
+        // Add one to the index because the select has the ROW__ID as the first column.
+        setColExprs.put(columnOffset + i, setCol);
+      }
+    }
+
+    sqlBuilder.append(" from ");
+    sqlBuilder.appendTargetTableName();
+
+    if (updateBlock.getWhereTree() != null) {
```

Comment:
isn't this redundant check since the above code doesn't do any checks when constructing `whereClause`

Reply:
👍